packages/drng/ticker.go

```diff
@@ -0,0 +1,104 @@
+package drng
+
+import (
+	"time"
+
+	"github.com/iotaledger/goshimmer/packages/clock"
+)
+
+const (
+	intervalCheckDRNG = 100 // interval for checking whether a dRNG message is received, in ms
+	checksPerSecond   = 1000 / intervalCheckDRNG
+)
+
+// Ticker holds a channel that delivers randomness at intervals.
+type Ticker struct {
+	dRNGState    func() *State
+	dRNGTicker   *time.Ticker
+	interval     int64 // the interval at which the ticker should tick (in seconds).
+	defaultValue float64
+	awaitOffset  int // defines the max amount of time (in seconds) to wait for the next dRNG round after the excected time has elapsed.
+	missingDRNG  bool
+	c            chan float64
+	exit         chan struct{}
+}
+
+// NewTicker returns a pointer to a new Ticker.
+func NewTicker(dRNGState func() *State, interval int64, defaultValue float64, awaitOffset int) *Ticker {
+	return &Ticker{
+		dRNGState:    dRNGState,
+		interval:     interval,
+		defaultValue: defaultValue,
+		awaitOffset:  awaitOffset,
+		missingDRNG:  true,
+		c:            make(chan float64),
+		exit:         make(chan struct{}),
+	}
+}
+
+// Start starts the Ticker.
+func (t *Ticker) Start() {
+	now := clock.SyncedTime().Unix()
+	nextTimePoint := ResolveNextTimePoint(now, t.interval)
+	time.AfterFunc(time.Duration(nextTimePoint-now)*time.Second, func() {
+		// send for the first time right after the timer is executed
+		t.send()
+
+		t.dRNGTicker = time.NewTicker(time.Duration(t.interval) * time.Second)
+		defer t.Stop()
+	out:
+		for {
+			select {
+			case <-t.dRNGTicker.C:
+				t.send()
+			case <-t.exit:
+				break out
+			}
+		}
+	})
+}
+
+// Stop stops the Ticker.
+func (t *Ticker) Stop() {
+	t.exit <- struct{}{}
+	t.dRNGTicker.Stop()
+}
+
+// C returns the channel from which random generated numbers can be consumed from.
+func (t *Ticker) C() <-chan float64 {
+	return t.c
+}
+
+// sends the next random number to the consumer channel.
+func (t *Ticker) send() {
+	randomness := t.defaultValue
+	if t.dRNGState() != nil {
+		// wait for next randomness from dRNG
+		for i := 0; i < t.awaitOffset*checksPerSecond; i++ {
+			if t.dRNGTicker != nil && t.missingDRNG && clock.Since(t.dRNGState().Randomness().Timestamp) < time.Duration(t.interval)*time.Second {
+				t.missingDRNG = false
+				timeToNextDRNG := t.dRNGState().Randomness().Timestamp.Add(time.Duration(t.interval) * time.Second).Sub(clock.SyncedTime())
+				t.dRNGTicker.Reset(timeToNextDRNG)
+			}
+			if clock.Since(t.dRNGState().Randomness().Timestamp) < time.Duration(t.awaitOffset)*time.Second {
+				randomness = t.dRNGState().Randomness().Float64()
+				if t.dRNGTicker != nil {
+					t.dRNGTicker.Reset(time.Duration(t.interval) * time.Second)
+				}
+				break
+			}
+			time.Sleep(intervalCheckDRNG * time.Millisecond)
+		}
+	}
```

Comment:
I came up with an idea to improve this waiting loop, but not sure if it's better or I was wrong about something.

here's how it works:
1. When drngTicker ticks, we first check if the randomness is fresh (`clock.Since(t.dRNGState().Randomness().Timestamp) < time.Duration(t.resolution)*time.Second`)
    * if yes, then we send out the randomness and reset drngTicker
    * if no, we reset drngTicker to `awaitOffset`
2. In the for loop, we wait for `awaitOffset`, listen to Randomness event to give us new randomness. If the clock ticks again (no fresh randomness), we use the default value

```suggestion
if t.dRNGState() != nil {
    // check if the randomess is "fresh"
    if t.dRNGTicker != nil && clock.Since(t.dRNGState().Randomness().Timestamp) < time.Duration(t.interval)*time.Second {
	randomness = t.dRNGState().Randomness().Float64()
	// the theoretical time that we should receive a new randomness
	timeToNextDRNG := t.dRNGState().Randomness().Timestamp.Add(time.Duration(t.interval) * time.Second).Sub(clock.SyncedTime())
	t.dRNGTicker.Reset(timeToNextDRNG)
    } else {
	// set ticker to awaitOffset
	t.dRNGTicker.Reset(time.Duration(t.awaitOffset) * time.Second)
    }

out:
    // if fresh randomness is not received, we wait for new randomness for awaitOffset seconds
    for {
	// abort if we already get the latest randomness
	if randomness != t.defaultValue {
		break out
	}

	select {
	    // receive randomness from Randomness event
	    case drngstate := <-t.fromRandamnessEvent:
		// check if the randomness is "fresh"
		if clock.Since(drngstate.Randomness().Timestamp) < time.Duration(t.awaitOffset)*time.Second {
			randomness = t.dRNGState().Randomness().Float64()
			if t.dRNGTicker != nil {
				t.dRNGTicker.Reset(time.Duration(t.interval) * time.Second)
			}
			break out
		}
	    case <-t.dRNGTicker.C:
		// still no new randomness within awaitOffset, take the default value, and reset dRNGTicker
		now := clock.SyncedTime().Unix()
		t.dRNGTicker.Reset(time.Duration(ResolveNextTimePoint(now, t.interval)-now) * time.Second)
		break out
	}
    }
}
```